tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc

```diff
@@ -0,0 +1,437 @@
+                              ", version=", serv_ver_major, ".", serv_ver_minor,
+                              ".", serv_ver_patch, "]");
+    } else {
+      return errors::Internal("Handshake Error [result=", handshake_res,
+                              ", version=", serv_ver_major, ".", serv_ver_minor,
+                              ".", serv_ver_patch, "]");
+    }
+  }
+
+  return Status::OK();
+}
+
+Status IgniteDatasetIterator::ScanQuery() {
+  TF_RETURN_IF_ERROR(client_->WriteInt(25));                   // Message length
+  TF_RETURN_IF_ERROR(client_->WriteShort(scan_query_opcode));  // Operation code
+  TF_RETURN_IF_ERROR(client_->WriteLong(0));                   // Request ID
+  TF_RETURN_IF_ERROR(
+      client_->WriteInt(JavaHashCode(cache_name_)));  // Cache name
+  TF_RETURN_IF_ERROR(client_->WriteByte(0));          // Flags
+  TF_RETURN_IF_ERROR(client_->WriteByte(null_val));   // Filter object
+  TF_RETURN_IF_ERROR(client_->WriteInt(page_size_));  // Cursor page size
+  TF_RETURN_IF_ERROR(client_->WriteInt(part_));       // part_ition to query
+  TF_RETURN_IF_ERROR(client_->WriteByte(local_));     // local_ flag
+
+  int64_t wait_start = std::chrono::duration_cast<std::chrono::milliseconds>(
+                           std::chrono::system_clock::now().time_since_epoch())
+                           .count();
+
+  int32_t res_len;
+  TF_RETURN_IF_ERROR(client_->ReadInt(&res_len));
+
+  int64_t wait_stop = std::chrono::duration_cast<std::chrono::milliseconds>(
+                          std::chrono::system_clock::now().time_since_epoch())
+                          .count();
+
+  LOG(INFO) << "Scan Query waited " << (wait_stop - wait_start) << " ms";
+
+  if (res_len < 12) return errors::Internal("Scan Query Response is corrupted");
+
+  int64_t req_id;
+  TF_RETURN_IF_ERROR(client_->ReadLong(&req_id));
+
+  int32_t status;
+  TF_RETURN_IF_ERROR(client_->ReadInt(&status));
+
+  if (status != 0) {
+    uint8_t err_msg_header;
+    TF_RETURN_IF_ERROR(client_->ReadByte(&err_msg_header));
+
+    if (err_msg_header == string_val) {
+      int32_t err_msg_length;
+      TF_RETURN_IF_ERROR(client_->ReadInt(&err_msg_length));
+
+      uint8_t* err_msg_c = new uint8_t[err_msg_length];
+      TF_RETURN_IF_ERROR(client_->ReadData(err_msg_c, err_msg_length));
+      std::string err_msg((char*)err_msg_c, err_msg_length);
+      delete[] err_msg_c;
+
+      return errors::Internal("Scan Query Error [status=", status, ", message=",
+                              err_msg, "]");
+    }
+    return errors::Internal("Scan Query Error [status=", status, "]");
+  }
+
+  TF_RETURN_IF_ERROR(client_->ReadLong(&cursor_id_));
+
+  LOG(INFO) << "Query Cursor " << cursor_id_ << " is opened";
+
+  int32_t row_cnt;
+  TF_RETURN_IF_ERROR(client_->ReadInt(&row_cnt));
+
+  remainder_ = res_len - 25;
+  page_ = std::unique_ptr<uint8_t>(new uint8_t[remainder_]);
+  ptr_ = page_.get();
+
+  int64_t start = std::chrono::duration_cast<std::chrono::milliseconds>(
```

Comment:
Optional: Use `tensorflow::Env::NowSeconds()` here and below.

Reply:
Thanks for advice. Replaced with `Env::Default()->NowMicros()`.